pkg/fqdn/helpers.go

```diff
@@ -59,17 +59,29 @@ func (n *NameManager) mapSelectorsToIPsLocked(fqdnSelectors sets.Set[api.FQDNSel
 
 			for name, ips := range lookupIPs {
 				if len(ips) > 0 {
-					log.WithFields(logrus.Fields{
-						"DNSName":      name,
-						"IPs":          ips,
-						"matchPattern": ToFQDN.MatchPattern,
-					}).Debug("Emitting matching DNS Name -> IPs for FQDNSelector")
-					ipsSelected = append(ipsSelected, ips...)
+					if log.Logger.IsLevelEnabled(logrus.DebugLevel) {
+						log.WithFields(logrus.Fields{
+							"DNSName":      name,
+							"IPs":          ips,
+							"matchPattern": ToFQDN.MatchPattern,
+						}).Debug("Emitting matching DNS Name -> IPs for FQDNSelector")
+					}
+					if ipsSelected == nil {
+						ipsSelected = ips
+					} else {
+						ipsSelected = append(ipsSelected, ips...)
+						dedup = true
+					}
 				}
 			}
 		}
+		if dedup {
+			s := make(sets.Set[netip.Addr], len(ipsSelected))
+			s.Insert(ipsSelected...)
+			ipsSelected = s.UnsortedList()
+		}
 
-		selectorIPMapping[ToFQDN] = ip.KeepUniqueAddrs(ipsSelected)
+		selectorIPMapping[ToFQDN] = ipsSelected
```

Comment:
Did you consider whether to just optimize `KeepUniqueAddrs()` itself this way? Or do you think that the ability to optionally skip in the case of known non-duplicated IPs is too important for the API that this cilium library function provides?

Reply:
So, there are three cases here:
- `ToFQDN.MatchName` is set. Then we never need to deduplicate, since we get a single set of IPs
- `ToFQDN.MatchPattern` is set, but only one name is returned. Again, we can skip deduplication
- `ToFQDN.MatchPattern` is set and multiple names are returned. Only then do we need to deduplicate.

The biggest win is by avoiding needless `ipsSelected = append(ipsSelected, ...)` calls, which cause heap allocations and CPU churn.

As for uniqueness, I whipped up a throwaway benchmark and found that `sets.Set` is faster for arrays under 100,000, and sorting is faster for arrays over 100,000. So, I think we should stick with `sets.Set` :-)

Reply:
Is there a reason not to just put the set logic into `ip.KeepUniqueAddrs()` to optimize all users?

(I'm fine with this PR as-is, we can always follow up on that separately if we think it's a good idea)